### PR TITLE
Issue #SB-16574: Rewriting tpd reports with analytics data product framework

### PR DESCRIPTION
**Pull request by ishawakankar**

### CODE REVIEW

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.ekstep.analytics.framework._
+import org.ekstep.analytics.util.CourseUtils.{getCourseBatchDetails, getLiveCourses, getTenantInfo, loadData}
+import org.sunbird.cloud.storage.conf.AppConf
+import org.ekstep.analytics.framework.util.CommonUtil
+
+case class BaseCourseMetricsOutput(courseName: String, batchName: String, status: String, slug: String, courseId: String, batchId: String) extends AlgoInput
+
+trait BaseCourseMetrics[T <: AnyRef, A <: BaseCourseMetricsOutput, B <: AlgoOutput, R <: Output] extends IBatchModelTemplate[T,BaseCourseMetricsOutput,B,R]{
+
+  override def preProcess(events: RDD[T], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[BaseCourseMetricsOutput] = {
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala]: Configuration is incorrect. It should take in the following parameters
- channelId - Array[String]
- courseId - Array[String]
- esQuery - String (Elastic search query)

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    val data = getCourseMetrics(config)
+    val encoder = Encoders.product[BaseCourseMetricsOutput]
+    val finalData = data.as[BaseCourseMetricsOutput](encoder).rdd
+    finalData.map(f => {
+      val batchStatus = if(f.status.equals("0")) "Upcoming" else if(f.status.equals("1")) "Ongoing" else if(f.status.equals("2")) "Expired" else ""
```
[the remark this answers is not in the document]
Comment (ishawakankar) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala]: Done.

File: platform-framework/analytics-core/src/main/scala/org/ekstep/analytics/framework/Models.scala [hunk 3]
```diff
@@ -242,4 +242,8 @@ case class DeviceProfileOutput(device_id: String, first_access: Option[Timestamp
                                state_custom: Option[String], state_code_custom: Option[String], district_custom: Option[String],
                                fcm_token: Option[String], producer_id: Option[String], user_declared_state: Option[String],
                                user_declared_district: Option[String], api_last_updated_on: Option[Timestamp], user_declared_on: Option[Timestamp],
-                               updated_date: Option[Timestamp] = Option(new Timestamp(System.currentTimeMillis()))) extends AlgoOutput
\ No newline at end of file
+                               updated_date: Option[Timestamp] = Option(new Timestamp(System.currentTimeMillis()))) extends AlgoOutput
+
+//Live Courses Information from druid: getLiveCourses
+case class CourseResult(pagingIdentifiers: Map[String, Int], dimensions: List[String], events: List[EventList])
+case class EventList(segmentId: String, offset: Int, event: Map[String,String])
```
[the remark this answers is not in the document]
Comment (ishawakankar) [reply platform-framework/analytics-core/src/main/scala/org/ekstep/analytics/framework/Models.scala]: Done.

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala (hunk 1 above)
Comment (ishawakankar) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala]: Done.

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+  override def preProcess(events: RDD[T], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[BaseCourseMetricsOutput] = {
+    val data = getCourseMetrics(config)
+    val encoder = Encoders.product[BaseCourseMetricsOutput]
+    val finalData = data.as[BaseCourseMetricsOutput](encoder).rdd
+    finalData.map(f => {
+      val batchStatus = BatchStatus.values.filter(c => (c.id==Integer.parseInt(f.status)))
+      if (batchStatus.isEmpty) f.copy(status = "") else f.copy(status = batchStatus.firstKey.toString)
+    })
+  }
```
[the remark this answers is not in the document]
Comment (utk14) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala]: Done. It is now being done in postprocess

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  def getCourseMetrics(config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): DataFrame = {
+    val liveCourses = CourseUtils.getCourse(config)
```
[the remark this answers is not in the document]
Comment (utk14) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/report/BaseCourseMetrics.scala]: Done

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/report/CourseEnrollmentModel.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      val reportConfig = JSONUtils.deserialize[ReportConfig](JSONUtils.serialize(configMap))
+
+      import sqlContext.implicits._
+      reportConfig.output.map { f =>
+        if (f.`type`.equals("csv")) {
+          val df = data.toDF().na.fill(0L)
+          CourseUtils.postDataToBlob(df, f,config)
+        } else {
+          val strData = data.map(f => JSONUtils.serialize(f))
+          AzureDispatcher.dispatch(strData.collect(), config)
```
[the remark this answers is not in the document]
Comment (utk14) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/report/CourseEnrollmentModel.scala]: We are using outputDispatcher now

**Comment by ishawakankar** [comment]
Closing the PR and redoing the changes from sunbird-core-dataproducts(release-2.8.0)

Link for new PR:
https://github.com/project-sunbird/sunbird-core-dataproducts/pull/9

**Pull request closed without merging**